cmd/aws-vpc-cni-init/main.go

```diff
@@ -0,0 +1,177 @@
+
+	log.Infof("Copying CNI plugin binaries ...")
+	err = cp.InstallBinaries(pluginBins, hostCNIBinPath)
+	if err != nil {
+		log.WithError(err).Errorf("Failed to install binaries")
+		return 1
+	}
+
+	log.Infof("Copied all CNI plugin binaries to %s\n", hostCNIBinPath)
+
+	var primaryMAC string
+	primaryMAC, err = imds.GetMetaData("mac")
+	if err != nil {
+		log.WithError(err).Fatalf("aws-vpc-cni init failed\n")
+		return 1
+	}
+
+	log.Infof("Found primaryMAC %s", primaryMAC)
+
+	links, err := netlink.LinkList()
+	if err != nil {
+		log.WithError(err).Fatalf("Failed to list links\n")
+		return 1
+	}
+
+	var primaryIF string
+	for _, link := range links {
+		if link.Attrs().HardwareAddr.String() == primaryMAC {
+			primaryIF = link.Attrs().Name
+			break
+		}
+	}
+
+	if primaryIF == "" {
+		log.Errorf("Failed to retrieve primary IF")
+		return 1
+	}
+	log.Infof("Found primaryIF %s", primaryIF)
+
+	sys := sysctl.New()
+	// Configure rp_filter in loose mode
+	entry := "net/ipv4/conf/" + primaryIF + "/rp_filter"
+	err = sys.SetSysctl(entry, 2)
+	if err != nil {
+		log.WithError(err).Fatalf("Failed to set rp_filter for %s\n", primaryIF)
+		return 1
+	}
+
+	val, _ := sys.GetSysctl(entry)
+	log.Infof("Updated %s to %d", entry, val)
+
+	// Enable or disable TCP early demux based on environment variable
+	disableIPv4EarlyDemux := getEnv(envDisableIPv4TcpEarlyDemux, "false")
+	entry = "net/ipv4/tcp_early_demux"
+	if disableIPv4EarlyDemux == "true" {
+		err = sys.SetSysctl(entry, 0)
+		if err != nil {
+			log.WithError(err).Fatalf("Failed to disable tcp_early_demux\n")
+			return 1
+		}
+	} else {
+		err = sys.SetSysctl(entry, 1)
+		if err != nil {
+			log.WithError(err).Fatalf("Failed to enable tcp_early_demux\n")
+			return 1
+		}
+	}
+
+	val, _ = sys.GetSysctl(entry)
+	log.Infof("Updated %s to %d", entry, val)
+
+	// Enable IPv6 when environment variable is set
+	// Note that IPv6 is not disabled when environment variable is unset. This is omitted to preserve default host semantics.
+	enableIPv6 := getEnv(envEnableIPv6, "false")
+	if enableIPv6 == "true" {
+		entry = "net/ipv6/conf/all/disable_ipv6"
+		err = sys.SetSysctl(entry, 0)
+		if err != nil {
+			log.WithError(err).Fatalf("Failed to set disable_ipv6 to 0\n")
+			return 1
+		}
+		val, _ = sys.GetSysctl(entry)
+		log.Infof("Updated %s to %d", entry, val)
+
+		entry = "net/ipv6/conf/all/forwarding"
+		err = sys.SetSysctl(entry, 1)
+		if err != nil {
+			log.WithError(err).Fatalf("Failed to enable ipv6 forwarding\n")
+			return 1
+		}
+		val, _ = sys.GetSysctl(entry)
+		log.Infof("Updated %s to %d", entry, val)
+
+		entry = "net/ipv6/conf/" + primaryIF + "/accept_ra"
+		err = sys.SetSysctl(entry, 2)
+		if err != nil {
+			log.WithError(err).Fatalf("Failed to enable ipv6 accept_ra\n")
+			return 1
+		}
+		val, _ = sys.GetSysctl(entry)
+		log.Infof("Updated %s to %d", entry, val)
+	}
+
+	// TODO: In order to speed up pod launch time, VPC CNI init container is not a Kubernetes init container.
```

Comment:
just a thought, why we make it two container and sync using filePath?
how about just one container, that runs the init first then rest? (if we want parallel processing, we can use goroutings)

Reply:
also nit:
since we are using golang, i prefer to split this "main" into smaller functions, such as `findNetLinkByMac`  `configureRPFilter()`, `configureIPv6Settings`

Reply:
That's a good point. Now that #2137 is removing the init container, we can actually just merge the two. I thought the init container was separate as it needed escalated privileges for setup, but it looks like that is not the case. I think we should do this in stages, though, i.e. this should merge with init container support, then #2137 can be rebased to merge with a single container.

Reply:
Did some more digging, and it looks like we cannot merge the two containers as the `init` container requires privileged  permissions to set `/proc/sys/net/ipv4/conf/eth0/rp_filter` to loose mode.

I see you had a suggestion in https://github.com/aws/amazon-vpc-cni-k8s/issues/796 to use `src_valid_mark` instead, so I am trying that now

Reply:
Oh nvm, setting that requires the same level of privilege, so yeah privilege is the limitation for the two containers

Reply:
In the latest diff, I broke up `aws-vpc-cni-init/main.go:main` into several parts. For `aws-vpc-cni/main.go:main`, I think it was already split up pretty well

Reply:
Ref for init-container use case - https://github.com/aws/amazon-vpc-cni-k8s/issues/796